library/src/main/java/com/novoda/downloadmanager/VersionOneToVersionTwoMigrator.java

```diff
@@ -21,35 +21,51 @@
     private final DownloadsPersistence downloadsPersistence;
     private final InternalFilePersistence internalFilePersistence;
     private final SqlDatabaseWrapper database;
-    private final Callback migrationCallback;
     private final UnlinkedDataRemover unlinkedDataRemover;
+    private final MigrationService migrationService;
+    private final NotificationChannelCreator channelCreator;
+    private final NotificationCreator<MigrationStatus> notificationCreator;
+    private final InternalMigrationStatus internalMigrationStatus;
 
     VersionOneToVersionTwoMigrator(MigrationExtractor migrationExtractor,
                                    DownloadsPersistence downloadsPersistence,
                                    InternalFilePersistence internalFilePersistence,
                                    SqlDatabaseWrapper database,
-                                   Callback migrationCallback,
-                                   UnlinkedDataRemover unlinkedDataRemover) {
+                                   UnlinkedDataRemover unlinkedDataRemover,
+                                   MigrationService migrationService,
+                                   NotificationChannelCreator channelCreator,
+                                   NotificationCreator<MigrationStatus> notificationCreator,
+                                   InternalMigrationStatus internalMigrationStatus) {
         this.migrationExtractor = migrationExtractor;
         this.downloadsPersistence = downloadsPersistence;
         this.internalFilePersistence = internalFilePersistence;
         this.database = database;
-        this.migrationCallback = migrationCallback;
         this.unlinkedDataRemover = unlinkedDataRemover;
+        this.migrationService = migrationService;
+        this.channelCreator = channelCreator;
+        this.notificationCreator = notificationCreator;
+        this.internalMigrationStatus = internalMigrationStatus;
     }
 
     @Override
     public void migrate() {
         unlinkedDataRemover.remove();
-        migrationCallback.onUpdate("Extracting Migrations");
+        internalMigrationStatus.markAsExtracting();
+        notifyOfProgress(internalMigrationStatus);
+
         Log.d(TAG, "about to extract migrations, time is " + System.nanoTime());
         List<Migration> migrations = migrationExtractor.extractMigrations();
         Log.d(TAG, "migrations are all EXTRACTED, time is " + System.nanoTime());
 
-        migrationCallback.onUpdate("Migrating Files");
+        internalMigrationStatus.markAsMigrating();
+        notifyOfProgress(internalMigrationStatus);
```

Comment:
I'm not sure if `marking` and then `notifying` adds to much noise in this class but it is super explicit that we are mutating an object (to prevent unnecessary memory churn) and then forwarding that for the `Service`.